Review REP 144

tfoote opened the pull request:
This is an pull-request to review and then ratify REP 144. A call for comments will be issued to ros-users@ after some time we'll call for a vote. Please use this PR for comments and voting when the time is appropriate. 

(For reference earlier discussions can be found in PR #94 )

mikepurvis commented:
Ah, this looks great. At Clearpath, we've standardized a few more package names across our platforms, including:

  - `xxx_desktop` (metapackage of stuff depending on rviz, rqt, but doesn't depend on drivers)
  - `xxx_viz` (launchers for rviz, especially `view_model.launch` and `view_robot.launch`)
  - `xxx_simulator` (metapackage of stuff depending on gazebo, but otherwise has no desktop dependencies)
  - `xxx_base` (nodes and driver dependencies which are part of the real robot, but shouldn't be installed on the desktop or by the simulator metapackage)
  - `xxx_control` (nodes/plugins/launchers which are common to simulated and real robot)

jbohren commented:
Can you elaborate on:

> third party libraries that are patched / integrated into ROS should not be named like their rosdep key as it creates a conflict across Ubuntu versions. If it is not specialized, name it generically `<name_of_library>_ros`

What about packages which are CMake only, and need to be built in order from source in a catkin workspace? In this case, creating package.xml files with the package's rosdep key as its package name makes everything build in the right order when using `catkin build` or `catkin_make_isolated `.

wjwwood commented:
> What about packages which are CMake only, and need to be built in order from source in a catkin workspace? In this case, creating package.xml files with the package's rosdep key as its package name makes everything build in the right order when using catkin build or catkin_make_isolated.

Yeah, that's convenient, but they can never be released like that because they will conflict with the rosdep key which can install them from the package manager.

If you are building something from source because the package manager does not have it yet then I believe it should have a different name than what the rosdep key would be. We had this issue with `pcl`, where he had a package `pcl` which was our release of `pcl` using a `package.xml` and our third party pipeline and the we wanted to switch to `pcl` from upstream. The issue was that rosdep keys are not ROS distro specific, so you cannot have a rosdep key in one version of ROS resolve to a ROS package and in another version of ROS have that same key resolve to a system dependency. So since we say that rosdep keys should try to match the Ubuntu key, and the rosdep key cannot overlap a ROS package name even across different distros, then you shouldn't name plain CMake projects with a `package.xml` the same as their apt-get key name.

You can always use the same name if you never release the plain CMake package.

mikepurvis commented:
Hrm. Maybe this REP should provide some specific guidance on this scenario.

For example, one of my colleagues has requested packaging a solver toolkit, and I could either just bloom third-party it, or I could set up debs in a launchpad PPA and then create a PR on reprepro-updater to pull them in, and on rosdistro to create a rosdep key for it. But it's not at all clear which is the preferred solution, and it seems like once you've decided, it's pretty painful to change your mind later.

If the REP could offer guidance into this and similar situations, that would be helpful. For example, how could the PCL situation have been handled differently to make the transition from ROS package to system package easier?

wjwwood commented:
The `pcl` example is an example of what worked (not a problem we had to solve), but I gave it as an example of why it matters. We named the new rosdep rules `libpcl-1.7{-dev}` which worked because the one we packaged ourselves was called `pcl`. However, if we had called the one we packaged `libpcl-1.7`, then creating a rosdep rule in the future would need a new name that didn't match Ubuntu's name. I'm just suggesting that when we roll our own versions of upstream dependencies that we do not release them with names that might collide in the future with rosdep keys. Based on this, I do not believe a third party package released with bloom should ever have the same name as you'd expect it to have in Ubuntu.

I think a recommendation of doing either a PPA or a third party bloom release is actually out of scope for this REP. So I would say that whether you release a third party package through bloom or use a PPA isn't really impacted by this guideline. You may choose to do one of those two things based on other considerations like the fact that a PPA only gets you Debian support, so you packages which depend on it will not work on RPM based systems and/or systems like OS X. But on the other hand putting it in a PPA and making a rosdep rule for it is probably a good solution if you expect it to be available in the different package managers in the future or if it is in Homebrew and/or RPM but not Debian. There are just a lot of different reasons you might choose to do one strategy over the other.

I suppose you could argue that the third party release REP could give pointers on this.

jack-oquin commented:
+1 to the overal proposal. It is very helpful.

> Ah, this looks great. At Clearpath, we've standardized a few more package names across our platforms, including:

> xxx_desktop (metapackage of stuff depending on rviz, rqt, but doesn't depend on drivers)
xxx_viz (launchers for rviz, especially view_model.launch and view_robot.launch)
xxx_simulator (metapackage of stuff depending on gazebo, but otherwise has no desktop dependencies)
xxx_base (nodes and driver dependencies which are part of the real robot, but shouldn't be installed on the desktop or by the simulator metapackage)
xxx_control (nodes/plugins/launchers which are common to simulated and real robot)

+1 to Mike's list of additional suffixes, and suggest most of them be included in this REP.

I not sure about the last two, but could probably be persuaded.

trainman419 commented:
It would be nice to give a few examples for the prefix rules. They're pretty complicated, particularly when more than one rule applies at the same time.

Are there any rules for robot-specific configurations for other packages; for example navigation and moveit?

mikepurvis commented:
http://wiki.ros.org/jackal_navigation
http://wiki.ros.org/husky_navigation
http://wiki.ros.org/cob_navigation

MoveIt has as a defacto standard the name chosen by the setup assistant:

http://www.ros.org/debbuild/indigo.html?q=_moveit_config

vrabaud commented:
ok, to summarize: I need to add a few examples and that's it.

I do agree with @mikepurvis and a few others on the original thread that we also need robot package naming conventions. Here, it's only about generic ROS package. Should I complete this one ? Can it be done later ? (we can update a REP so why not). Should it be a different REP ?
It is definitely useful but also pretty hard: what is a navigation package for a robot ? Does it need to have specific services / messages ?
You can read https://github.com/ros-infrastructure/rep/pull/94 for insight, especially the ROS industrial references:
https://github.com/gavanderhoorn/rep/blob/repository_layout/rep-xxxx.rst
http://wiki.ros.org/Industrial/Tutorials/SuggestedPackageLayoutNewRepositories

tfoote commented:
@vrabaud I suggest we confirm this and can have another conversation about the robot packages. It may end up with updating this REP or making a new one. But we should ratify what we agree upon before we approach the more complicated topic.

vrabaud commented:
before merging that, we should finish https://github.com/ros-infrastructure/rep/pull/105
Please comment there for any examples you'd like to add.

tfoote commented:
I merged in #105  It contained the noteable change of moving from our (unstated) standard `test_*` to `*_tests` as the test package naming guide.

We've let this sit for quite a while does anyone object to ratifying it in the current form?

jack-oquin commented:
+1 LGTM

wjwwood commented:
Under mandatory rules it mentions use of `_`, but it doesn't say anything about avoiding double underscore (`__`). It would be nice to get this point in, since we often use this to give us a "safe" separator when combining the package name with some other suffix or prefix in generated code. For instance, if a package called `foo` produces two symbols `foo` and `foo_msg` then we have no protection against another package being called `foo_msg` which produces `foo_msg` and `foo_msg_msg`. Having a `__` be an illegal segment in the package name allows us to make `foo` give `foo` and `foo__msgs` which cannot be accidentally collided with.

Other than that point, I'd say +1. We should also solicit voting from ros-users@ if we haven't already.

tfoote commented:
From @wjwwood I added a statement that you cannot use consecutive `_` separators.

dirk-thomas commented:
The `Mandatory Rules` are pretty confusing and the rational is not always correct. I will try to rephrase the paragraph when I find time for it.

A brain dump of things which imo should be updated (if anybody want to go ahead and change it):
* the first two bullets are not rules but definitions of terms
* the rational for underscore is wrong - it has nothing do to with OS conventions
* the rational for repeated underscores does not mention what generated symbols it refers to making it hard to understand what it actually means
* the rational for the first character being alphabetic is wrong
* explicitly mention case (must be lower case)

dirk-thomas commented:
@tfoote What is the status of this PR?

tfoote commented:
It's waiting for the full form of your feedback @dirk-thomas before we circulate it broadly for ratification.

dirk-thomas commented:
I am sorry but I won't have time to spend on rephrasing the paragraph (especially since I neither wrote the original draft nor created this PR). Please consider using my feedback from the bullet points to update the patch if you think they should be applied.

tfoote commented:
@dirk-thomas Please review the updated mandatory rules section

dirk-thomas commented on rep-0144.rst:
```diff
@@ -31,42 +31,55 @@ Some of those rules are mandatory, others merely advised.
-  make it hard to follow OSes conventions/
-* they must be at least two characters long and must start with an alphabetic character.
+
+For clarity these are both specifically exclusively lowercase.
```
I don't think "clarity" is a good reason for this. Namespaces usually map to directory names and those might not support case sensitive names depending on the filesystem. Since names which only differ in case would collide in that scenario lower case is is being used exclusively.

tfoote replied:
This statement isn't giving a reason, it's clarifying that `a-z` is specified specifically to not include `A-Z`. I can add the additional reason for the lowercase in the rules below.

davetcoleman commented:
this PR is 3 years old!

peci1 commented:
+1

SteveMacenski commented:
+1
This all seems common sense and generally as most folks have been operating for awhile.

DLu commented:
+1. 

Fun fact: there are three 2 letter packages: [ff](https://jsk-docs.readthedocs.io/en/latest/jsk_planning/doc/pddl/pddl_planner/README.html) [tf](https://github.com/tfoote) and [mk](https://github.com/ros/ros/tree/indigo-devel/core/mk)

Here are the three letter packages: bfl, csm, ecl, eml, enu, fcl, gsd, joy, kni, kvh, lkh, msp, ocl, pid, rdl, ros, rqt, rtt, viz, xpp

mintar commented:
+1

ross-desmond commented:
+1

tfoote commented:
Having addressed the remaining comments and with several +1's i'm going to merge this finally. Thanks everyone for the feedback and sticking with us on this.
